### PR TITLE
Block SetWarmUp after Algorithm has compeleted initialization

### LINKED ISSUE #4939: SetWarmup Throws Exception When Not Called in Initialize
#### Expected Behavior
Notify users that `QCAlgorithm.SetWarmup` cannot be used after initialization. 

#### Actual Behavior
`QCAlgorithm.SetWarmup` can be called after initialization which doesn't do any warm up.

#### Potential Solution
Throw exception if `SetWarmup` is called when `IsWarmingUp` has already been flipped to `false`. The message should tell the user that this method should be called in `Initialize` only.

#### Reproducing the Problem
Call `SetWarmup` in `OnData`.

#### Checklist
- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

### PR DESCRIPTION
**Pull request by C-SELLERS**
#### Description

Now we throw an error when SetWarmUp is called after Algorithm Initialization

#### Related Issue

Closes #4939

#### Motivation and Context

Shouldn't be able to mess with warmup settings after warmup has already occurred. 

#### Requires Documentation Change

N/A

#### How Has This Been Tested?

All previous regressions pass, included new regression that intentionally throws and is checked for runtime error status.

#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Algorithm/QCAlgorithm.History.cs [hunk 1]
```diff
@@ -51,6 +51,11 @@ public bool IsWarmingUp
         /// <param name="timeSpan">The amount of time to warm up, this does not take into account market hours/weekends</param>
         public void SetWarmup(TimeSpan timeSpan)
         {
+            if (!IsWarmingUp)
```
Comment (Martin-Molinero) [inline Algorithm/QCAlgorithm.History.cs:54]: Cool!
I think more precisely we could use the algorithms `_locked` field, which is false until after being initialized, seems more exact for the behavior we want? Think calling `SetWarmup` during warmup should also throw.

Maybe we could make this method directly call `SetWarmup(TimeSpan timeSpan, Resolution resolution)`. With resolution null. Same for `public void SetWarmup(int barCount)` -> `public void SetWarmup(int barCount, Resolution resolution)`

File: Algorithm.CSharp/WarmUpAfterIntializeRegression.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            SetStartDate(2013, 10, 07);  //Set Start Date
+            SetEndDate(2013, 10, 11);    //Set End Date
+            SetCash(100000);
+            var equity = AddEquity("SPY");
+        }
+
+        public override void OnData(Slice slice)
+        {
+            // Should throw and set Algorithm status to be runtime error
+            SetWarmUp(TimeSpan.FromDays(2));
```
Comment (Martin-Molinero) [inline Algorithm.CSharp/WarmUpAfterIntializeRegression.cs:41]: Nice 👍

File: Tests/RegressionTests.cs [hunk 3]
```diff
@@ -77,7 +77,8 @@ private static TestCaseData[] GetRegressionTestParameters()
             var nonDefaultStatuses = new Dictionary<string, AlgorithmStatus>
             {
                 {"TrainingInitializeRegressionAlgorithm", AlgorithmStatus.RuntimeError},
-                {"OnOrderEventExceptionRegression", AlgorithmStatus.RuntimeError}
+                {"OnOrderEventExceptionRegression", AlgorithmStatus.RuntimeError},
+                {"WarmUpAfterIntializeRegression", AlgorithmStatus.RuntimeError }
```
Comment (Martin-Molinero) [inline Tests/RegressionTests.cs:81]: A separate issue, but starting to think this end state should probably be owned by the algorithm itself, not this 3rd party 🤔

Comment (C-SELLERS) [reply Tests/RegressionTests.cs:81]: I could see that, maybe we should consider extending the IRegressionAlgorithmDefinition interface to contain a member for expected status?

File: Algorithm/QCAlgorithm.History.cs (hunk 1 above)
Comment (C-SELLERS) [reply Algorithm/QCAlgorithm.History.cs:54]: Gotcha!

**Review by Martin-Molinero — APPROVED** [review]

**Pull request merged**